src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,652 @@
+ * omitted if the KDF algorithm has a fixed or default PRF or if the PRF is
+ * specified as part of the {@code KDFParameters}.
+ * <p>
+ * The class has two derive methods, {@code deriveKey} and {@code deriveData}.
+ * The {@code deriveKey} method accepts an algorithm {@code String} and
+ * will return a {@code SecretKey} object with the specified algorithm. The
+ * {@code deriveData} method returns a byte array of raw data.
+ * <p>
+ * If a provider is not specified in the {@code getInstance} method when
+ * instantiating a {@code KDF} object, the provider is selected the first time
+ * the {@code deriveKey} or {@code deriveData} method is called, and a provider
+ * is chosen that supports the parameters passed to the {@code deriveKey} or
+ * {@code deriveData} method. However, if {@code getProviderName} or
+ * {@code getKDFParameters} is called before calling the {@code deriveKey} or
+ * {@code deriveData} methods, the first provider supporting the KDF algorithm
+ * and {@code KDFParameters} is chosen which may not be the provider that is
+ * eventually selected once the {@code AlgorithmParameterSpec} is supplied in
+ * the derive methods; therefore it is recommended not to call
+ * {@code getProviderName} or {@code getKDFParameters} until after a key
+ * derivation operation. Once a provider is selected, it cannot be changed.
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDFWithHmacSHA256");
+ *
+ *    AlgorithmParameterSpec derivationParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *    kdfHkdf.deriveKey("AES", derivationParameterSpec);
+ *}
+ *
+ * @see KDFParameters
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    int DERIVE_KEY = 0;
+    int DERIVE_DATA = 1;
```

Comment:
These two variables are a waste of space... `chooseProvider()` impl can just use a local boolean variable to remember whether the derivation is for key or for data.

Reply:
Yes, resolved.